### PR TITLE
V3 0 0 component rendering into gitbook and docx formats

### PR DESCRIPTION
**Pull request by jcscottiii**
This patch will introduce ways to render the version 3.0.0 components which were introduced in #160.

For Gitbook:

- The Narrative Key is printed as a header
- The Narrative Text is printed afterwards under the narrative key.
- Responsible Role is printed after the Component Name.

For Docx:

There are new templating functions:

- `{{getResponsibleRole "standard" "control"}}` - Will get the responsible role.
- `{{getAllControlSections “standard” "control”}}` - Will get all the narratives for a control
- `{{getControlSection “standard” “control” “section key”}}` - Will get the narrative for that given section of that control.
- {{getParameter “standard” “control” “parameter key”}}` - Will get the parameter for that given parameter key of that control.

Docx Template:
image

Docx Filled Template:
image

### CODE REVIEW

File: docx/docx.go [hunk 1]
```diff
@@ -23,66 +24,156 @@ type OpenControlDocx struct {
...
+	}
 	docTemplate.AddFunctions(funcMap)
 	docTemplate.Parse()
 	docTemplate.Execute(config.ExportPath, nil)
 	return err
 }
 
-// FormatControl returns a control formatted for docx
-func (openControl *OpenControlDocx) FormatControl(standardControl string) string {
+type componentInfoType int
```
Comment (afeld) [inline docx/docx.go]: Is there a point to making the sub-type here? (legitimately asking)

File: docx/docx.go [hunk 2]
```diff
@@ -23,66 +24,156 @@ type OpenControlDocx struct {
...
+
+			// If section header exists, let's print it's corresponding text and not the header itself.
+			if specifiedSections.Has(section.GetKey()) {
+				text = fmt.Sprintf("%s%s\n", text, section.GetText())
+				foundText = true
+			}
+		}
+	}
+	return text, foundText
+}
```
Comment (afeld) [inline docx/docx.go]: This function is doing a lot...can we split it up? Example functions that could come out of it:

* `getCombinedControlInfo()`
* `getSectionInfo()`
* `getNarrative()`

Also, not sure that `foundText` is necessary, but it's not hurting anything I suppose.

File: docx/docx.go [hunk 3]
```diff
@@ -23,66 +24,156 @@ type OpenControlDocx struct {
...
+
+	for _, parameter := range justification.SatisfiesData.GetParameters() {
+		// If section header exists, let's print it's corresponding text and not the header itself.
+		if specifiedSections.Has(parameter.GetKey()){
+			text = fmt.Sprintf("%s%s\n", text, parameter.GetText())
+			foundText = true
+		}
+	}
+	return text, foundText
+}
```
Comment (afeld) [inline docx/docx.go]: Mind extracting a `getParameter()` function?

File: docx/docx.go [hunk 4]
```diff
@@ -23,66 +24,156 @@ type OpenControlDocx struct {
...
-						verification := component.GetVerifications().Get(coveredBy.VerificationKey)
-						text += fmt.Sprintf("- %s %s  \n", verification.Name, verification.Path)
-					}
-				})
-			}
+				if found {
+					text = fmt.Sprintf("%s%s", text, specificText)
+				} else {
+					text = fmt.Sprintf("%s%s%s\n", text, specificText, constants.WarningNoInformationAvailable)
+				}
```
Comment (afeld) [inline docx/docx.go]: Can we extract this to a method?

File: gitbook/gitbookCertification.go [hunk 5]
```diff
@@ -13,10 +13,31 @@ func (openControl *OpenControlGitBook) exportControl(control *ControlGitbook) (s
...
+					text = fmt.Sprintf("%sNo narrative found for the combination of standard %s and control %s\n", text, control.standardKey, control.controlKey)
+					return
+				}
+
+				for _, narrative := range justification.SatisfiesData.GetNarratives() {
+					if narrative.GetKey() != "" {
+						text = fmt.Sprintf("%s\n##### %s\n", text, narrative.GetKey())
+					}
+					text = fmt.Sprintf("%s%s\n", text, narrative.GetText())
+				}
```
Comment (afeld) [inline gitbook/gitbookCertification.go]: Can we move everything/most from this callback into a function?

Comment (afeld) [inline gitbook/gitbookCertification.go]: Ditto with the outer callback. Basically, break `exportControl()` into smaller pieces.

File: docx/docx.go (hunk 1 above)
Comment (jcscottiii) [inline docx/docx.go]: yeah. I want so create some type of enumeration of what type of data i want going into the common function `getComponentText`. i can't pass a int by itself, only through those value declared by my new type.

**Comment by afeld** [comment]
Merge from `master` (so the PR updates) when ready!

File: docx/docx_test.go [hunk 6]
```diff
@@ -1,55 +1,104 @@
-package docx_test
+package docx
```
Comment (afeld) [inline docx/docx_test.go]: Is this right?

Comment (jcscottiii) [inline docx/docx_test.go]: yeah. So when you split out your test to the `packagename_test` (in this case `docx_test`) package, you can only test exported functions.

However, there was some edge case here https://github.com/opencontrol/compliance-masonry/pull/161/files#diff-13ce5f19e770df462c7202a0913535e5R80
that I didn't have a fixture for. So i needed access to an internal method so i have to move the test to the `packagename` (in this case `docx`) package.

Comment (afeld) [inline docx/docx_test.go]: I'd add an inline comment about that, in that case. After that, 🚢 !
